include/heuristic/HeuristicMapper.hpp

```diff
@@ -49,125 +49,101 @@ class HeuristicMapper : public Mapper {
      * (further layers contribute less) */
     double lookaheadPenalty = 0.;
-    /** number of swaps used to get from mapping after last layer to the current
-     * mapping */
-    std::size_t nswaps = 0;
+    /** number of swaps that were shared with another considered qubit such
+     * that both qubits got closer to being validly mapped*/
+    std::size_t sharedSwaps = 0;
     /** depth in search tree (starting with 0 at the root) */
     std::size_t depth  = 0;
     std::size_t parent = 0;
-    std::size_t id;
+    std::size_t id     = 0;
     /** true if all qubit pairs are mapped next to each other on the
      * architecture */
-    bool done = true;
-    /** controls if fidelity-aware heuristic should be used */
-    bool considerFidelity = false;
-    /** controls if admissible heuristic should be used */
-    bool admissibleHeuristic = true;
-
-    explicit Node(std::size_t nodeId, const bool considerFid = false,
-                  const bool admissibleHeur = true)
-        : id(nodeId), considerFidelity(considerFid),
-          admissibleHeuristic(admissibleHeur){};
+    bool validMapping = true;
+
+    explicit Node() {
+      qubits.fill(DEFAULT_POSITION);
+      locations.fill(DEFAULT_POSITION);
+    };
+    explicit Node(std::size_t nodeId) : id(nodeId) {
+      qubits.fill(DEFAULT_POSITION);
+      locations.fill(DEFAULT_POSITION);
+    };
     Node(std::size_t nodeId, std::size_t parentId,
          const std::array<std::int16_t, MAX_DEVICE_QUBITS>& q,
          const std::array<std::int16_t, MAX_DEVICE_QUBITS>& loc,
-         const std::vector<std::vector<Exchange>>&          sw = {},
-         const double initCostFixed = 0, const std::size_t searchDepth = 0,
-         const bool considerFid = false, const bool admissibleHeur = true)
-        : costFixed(initCostFixed), depth(searchDepth), parent(parentId),
-          id(nodeId), considerFidelity(considerFid),
-          admissibleHeuristic(admissibleHeur) {
+         const std::vector<Exchange>&                       sw            = {},
+         const std::set<Edge>&                              valid2QGates  = {},
+         const double                                       initCostFixed = 0,
+         const double      initCostFixedReversals                         = 0,
+         const std::size_t searchDepth                                    = 0,
+         const std::size_t initSharedSwaps                                = 0)
+        : costFixed(initCostFixed), costFixedReversals(initCostFixedReversals),
+          sharedSwaps(initSharedSwaps), depth(searchDepth), parent(parentId),
+          id(nodeId) {
       std::copy(q.begin(), q.end(), qubits.begin());
       std::copy(loc.begin(), loc.end(), locations.begin());
       std::copy(sw.begin(), sw.end(), std::back_inserter(swaps));
+      std::copy(valid2QGates.begin(), valid2QGates.end(),
+                std::inserter(validMappedTwoQubitGates,
+                              validMappedTwoQubitGates.begin()));
```

Comment:
I guess this is one of those points where I have lost the original intention behind the code. Why are these 4 parameters here copied explicitly with `std::copy` and not just using the regular copy constructor?

Reply:
That's right. I just followed the pre-existing lines above without much thought. I'm not too familiar with how C++ handles struct constructors in this regard, but don't we save an unnecessary vector allocation this way, since the copy constructor would overwrite the default initialization? Or does the default initialization only happen if there is no other initialization in the constructor?

Reply:
I am fairly sure that you are right when no optimizations are involved, but I think that modern compilers are clever enough to optimize any unnecessary initialisations away.
I guess you could just replace the existing code and see if anything breaks 🤷🏼

Reply:
Alright, I looked into this in more detail:

For vector members:
1. A simple copy assignment in the constructor body, does actually result in 2 initializations (the default initialization and a reinitialization), which also seems to not get optimized away by GCC (looking at the assembly of a minimal example)
2. By using the initialization list of the constructor the default initialization is skipped and the vector is directly initialized to the correct size, resulting in only 1 initialization.
3. `std::copy` of course also results in 1 initialization (the default initialization), but the vector might get resized during copying

For array members all 3 methods are equivalent copying just the data over, since the arrays are allocated within the struct.

So in theory 2. (initialization list) should be the fastest, then 3. (`std::copy`) and then 1. (copy assignment)

I ran a few tests with QMAP comparing the runtime (averaging multiple runs, which were executed in parallel with varying starting order to reduce effects of environment and scheduler).
I have _no_ idea why, but the order of their mapping times is exactly reversed to what I expected. The initialization list is about 1% slower than the current solution, and copy assignments in the constructor body are about 0.5% faster. These differences aren't huge but I ran the experiment often enough, that the results are statistically significant.

So should we switch to copy assignments (in that case we probably need to disable the linter warning to use the initialization list on those lines) or leave it as is? Or do we take the 1% performance decrease for slightly cleaner code?


Btw, while at it, I also tested the effects of using vectors instead of arrays for `qubits` and `locations` (both in the mapper and the nodes; using fixed size vectors with equal length to the number of output qubits), since the current solution of hard-coded 128-length arrays is both pretty wasteful space-wise and not too future-proof. Unfortunately, this results in a significant performance decrease of about 40%.

Reply:
Thanks for the thorough testing. 
I guess I would just take the 1% performance hit to have the cleaner code. That 1% won't really be noticeable in real world examples.

It's really unfortunate that the change from arrays to vectors is so costly. But maybe we'll find a way to circumvent that somehow in the future.